lib/self-evict.js

```diff
@@ -179,10 +180,68 @@ SelfEvict.prototype._preEvict = function _preEvict() {
 };
 
 SelfEvict.prototype._evict = function _evict() {
+    var self = this;
+
     this._transitionTo(PhaseNames.Evicting);
 
-    this.ringpop.membership.setLocalStatus(Member.Status.faulty);
-    process.nextTick(this._postEvict.bind(this));
+    var pingEnabled = this.ringpop.config.get('selfEvictionPingEnabled');
+
+    var restartGossip = false;
+    // Stop gossip when currently gossiping and ping is enabled
+    if (!this.ringpop.gossip.isStopped && pingEnabled) {
+        this.ringpop.gossip.stop();
+        restartGossip = true;
+    }
+
+    if (pingEnabled) {
+        waitUntilNotPinging();
+
+        function waitUntilNotPinging() {
+            if (self.ringpop.gossip.isPinging) {
+                //Wait until isPinging is false...
+                setTimeout(waitUntilNotPinging, 25);
+            } else {
+                self.ringpop.membership.setLocalStatus(Member.Status.faulty);
+                gossipEviction();
+            }
+        }
+
+        function gossipEviction() {
+            var selfEvictionMaxPingRatio = self.ringpop.config.get('selfEvictionMaxPingRatio');
+            var membership = self.ringpop.membership;
+
+            var numberOfPingableMembers = _.filter(membership.members, membership.isPingable.bind(membership)).length;
+            var maxNumberOfTicks = Math.ceil(numberOfPingableMembers * selfEvictionMaxPingRatio);
+
+            // Never ping more than the maximum piggy back counter.
+            var numberOfPings = Math.min(
+                self.ringpop.dissemination.maxPiggybackCount,
+                maxNumberOfTicks
+            );
+
+            tick();
+
+            function tick() {
+                if (numberOfPings > 0) {
+                    numberOfPings--;
+                    self.ringpop.gossip.tick(tick);
```

Comment:
`gossip.tick()` executes a `ping` and when it fails it fans out a `pingReq`, I wonder if we want to wait for ping timeouts and then a pingreq which involves a timeout as well. I think only doing a ping is enough for the shutting down phase.